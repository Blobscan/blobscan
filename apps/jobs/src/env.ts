import { booleanSchema, createEnv, presetEnvOptions, z } from "@blobscan/zod";

export const env = createEnv({
  envOptions: {
    server: {
      DATABASE_URL: z.string().url(),
      // ETH Price is retrieved every second from the Chainlink: ETH/USD oracle
      // in the Polygon network.
      ETH_PRICE_SYNCER_ENABLED: booleanSchema.default("false"),
      ETH_PRICE_SYNCER_CRON_PATTERN: z.string().default("* * * * *"),
      ETH_PRICE_SYNCER_CHAIN_ID: z.coerce.number().default(137),
      ETH_PRICE_SYNCER_CHAIN_JSON_RPC_URL: z.string().url().optional(),
      ETH_PRICE_SYNCER_ETH_USD_PRICE_FEED_CONTRACT_ADDRESS: z
        .string()
        .default("0xF9680D99D6C9589e2a93a78A04A279e509205945"),
      ETH_PRICE_SYNCER_TIME_TOLERANCE: z.coerce.number().positive().optional(),

      REDIS_URI: z.string().default("redis://localhost:6379"),
    },
    ...presetEnvOptions,
  },
});
