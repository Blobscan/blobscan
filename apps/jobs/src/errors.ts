export class ErrorException extends Error {
  constructor(message: string, cause?: unknown) {
    super(message, {
      cause,
    });

    this.name = this.constructor.name;
  }
}
