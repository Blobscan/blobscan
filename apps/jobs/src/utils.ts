export type Granularity = "minute" | "hour" | "day";

export function determineGranularity(cronPattern: string): Granularity {
  switch (cronPattern) {
    case "0 * * * *":
      return "hour";
    case "0 0 * * *":
      return "day";
    case "* * * * *":
      return "minute";
    default:
      throw new Error(`Unsupported cron pattern: ${cronPattern}`);
  }
}

export function gracefulShutdown(closeOp: () => void | Promise<void>) {
  const shutdown = async (signal: string) => {
    console.log(`Received ${signal}`);

    await closeOp();

    process.exit(0);
  };

  process.on("SIGINT", () => void shutdown("SIGINT"));
  process.on("SIGTERM", () => void shutdown("SIGTERM"));
}
