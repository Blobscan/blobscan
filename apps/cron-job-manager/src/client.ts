import { createTRPCProxyClient, httpBatchLink } from "@trpc/client";
import superjson from "superjson";

import { type AppRouter } from "@blobscan/api";

import { BLOBSCAN_API_ENDPOINT } from "./env";

export const client = createTRPCProxyClient<AppRouter>({
  links: [
    httpBatchLink({
      url: `${BLOBSCAN_API_ENDPOINT}/api/trpc`,
    }),
  ],
  transformer: superjson,
});
