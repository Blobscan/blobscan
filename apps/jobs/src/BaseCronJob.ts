/* eslint-disable @typescript-eslint/no-misused-promises */
import { Queue, Worker } from "bullmq";
import type { Redis } from "ioredis";

import { createModuleLogger } from "@blobscan/logger";
import type { Logger } from "@blobscan/logger";

import { ErrorException } from "./errors";
import { createRedis } from "./redis";

export interface CommonCronJobConfig {
  redisUriOrConnection: Redis | string;
  cronPattern: string;
}

export interface BaseCronJobConfig extends CommonCronJobConfig {
  name: string;
  jobFn: () => Promise<void>;
}

export class CronJobError extends ErrorException {
  constructor(cronJobName: string, message: string, cause: unknown) {
    super(`Cron job "${cronJobName}" failed: ${message}`, cause);
  }
}

export class BaseCronJob {
  name: string;
  cronPattern: string;

  protected jobFn: () => Promise<void>;
  protected logger: Logger;

  protected connection: Redis;
  protected worker: Worker | undefined;
  protected queue: Queue | undefined;

  constructor({
    name,
    cronPattern,
    redisUriOrConnection,
    jobFn,
  }: BaseCronJobConfig) {
    this.name = `${name}-cron-job`;
    this.cronPattern = cronPattern;
    this.logger = createModuleLogger(this.name);

    let connection: Redis;

    if (typeof redisUriOrConnection === "string") {
      connection = createRedis(redisUriOrConnection);
    } else {
      connection = redisUriOrConnection;
    }

    this.queue = new Queue(this.name, {
      connection,
    });

    this.worker = new Worker(this.queue.name, jobFn, {
      connection,
    });

    this.queue.on("error", (err) => {
      this.logger.error(new ErrorException("A queue error occurred", err));
    });

    this.worker.on("failed", (_, err) => {
      this.logger.error(new ErrorException("A worker error ocurred", err));
    });

    this.connection = connection;
    this.jobFn = jobFn;
  }

  async start() {
    try {
      const jobName = `${this.name}-job`;
      const repeatableJob = await this.queue?.add(jobName, null, {
        repeat: {
          pattern: this.cronPattern,
        },
      });

      this.logger.debug("Cron job started successfully");

      return repeatableJob;
    } catch (err) {
      throw new CronJobError(
        this.name,
        "An error ocurred when starting cron job",
        err
      );
    }
  }

  close() {
    const teardownPromise: Promise<void> = Promise.resolve();

    return teardownPromise
      .finally(async () => {
        await this.#performClosingOperation(() =>
          this.worker?.removeAllListeners().close(true)
        );
      })
      .finally(async () => {
        await this.#performClosingOperation(() =>
          this.queue?.obliterate({ force: true })
        );
      })
      .finally(async () => {
        await this.#performClosingOperation(() =>
          this.queue?.removeAllListeners().close()
        );

        this.logger.info("Cron job closed successfully");
      });
  }

  async #performClosingOperation(operation: () => Promise<void> | undefined) {
    try {
      await operation();
    } catch (err) {
      const err_ = new CronJobError(
        this.name,
        "An error ocurred when performing closing operation",
        err
      );

      throw err_;
    }
  }
}
