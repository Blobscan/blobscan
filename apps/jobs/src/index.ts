import { logger } from "@blobscan/logger";

import type { BaseCronJob } from "./cron-jobs/BaseCronJob";
import { DailyStatsCronJob } from "./cron-jobs/daily-stats/DailyStatsCronJob";
import { EthPriceCronJob } from "./cron-jobs/eth-price/EthPriceCronJob";
import { OverallStatsCronJob } from "./cron-jobs/overall-stats/OverallStatsCronJob";
import { SwarmStampCronJob } from "./cron-jobs/swarm-stamp/SwarmStampCronJob";
import { env } from "./env";
import { createRedis } from "./redis";
import { getNetworkDencunForkSlot, gracefulShutdown } from "./utils";

const cronJobs: BaseCronJob[] = [];

async function main() {
  env.display();

  const redis = createRedis(env.REDIS_URI);

  cronJobs.push(
    new DailyStatsCronJob({
      cronPattern: env.STATS_SYNCER_DAILY_CRON_PATTERN,
      redisUriOrConnection: redis,
    }),
    new OverallStatsCronJob({
      cronPattern: env.STATS_SYNCER_OVERALL_CRON_PATTERN,
      redisUriOrConnection: redis,
      forkSlot:
        env.DENCUN_FORK_SLOT ?? getNetworkDencunForkSlot(env.NETWORK_NAME),
    })
  );

  if (env.ETH_PRICE_SYNCER_ENABLED) {
    if (!env.ETH_PRICE_SYNCER_CHAIN_JSON_RPC_URL) {
      logger.warn(
        `Skipping eth price cron job: ETH_PRICE_SYNCER_CHAIN_JSON_RPC_URL not defined`
      );
    } else {
      cronJobs.push(
        new EthPriceCronJob({
          cronPattern: env.ETH_PRICE_SYNCER_CRON_PATTERN,
          redisUriOrConnection: redis,
        })
      );
    }
  }

  if (env.SWARM_STAMP_SYNCER_ENABLED) {
    if (!env.BEE_ENDPOINT) {
      logger.warn(`Skipping swarm stamp cron job: BEE_ENDPOINT not defined`);
    } else {
      cronJobs.push(
        new SwarmStampCronJob({
          beeEndpoint: env.BEE_ENDPOINT,
          batchId: env.SWARM_BATCH_ID,
          cronPattern: env.SWARM_STAMP_CRON_PATTERN,
          redisUriOrConnection: redis,
        })
      );
    }
  }

  await Promise.all(cronJobs.map((j) => j.start()));
}

async function shutdown() {
  for (const job of cronJobs) {
    try {
      await job.close();
    } catch (err) {
      logger.error(`Error closing job ${job.constructor.name}:`, err);
    }
  }
}

main().catch(async (err) => {
  console.error(err);

  await shutdown();

  process.exit(0);
});

gracefulShutdown(shutdown);
