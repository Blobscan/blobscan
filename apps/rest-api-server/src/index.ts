/* eslint-disable @typescript-eslint/no-misused-promises */

import * as Sentry from "@sentry/node";
import bodyParser from "body-parser";
import cors from "cors";
import express from "express";
import morgan from "morgan";
import swaggerUi from "swagger-ui-express";
import { createOpenApiExpressMiddleware } from "trpc-openapi";

import {
  appRouter,
  createTRPCContext,
  metricsHandler,
  gracefulShutdown as apiGracefulShutdown,
} from "@blobscan/api";
import { logger } from "@blobscan/logger";
import { collectDefaultMetrics } from "@blobscan/open-telemetry";
import { StatsSyncer } from "@blobscan/stats-syncer";

import { env } from "./env";
import { openApiDocument } from "./openapi";
import { getNetworkDencunForkSlot } from "./utils";

collectDefaultMetrics();

const statsSyncer = new StatsSyncer({
  redisUri: env.REDIS_URI,
  lowestSlot:
    env.DENCUN_FORK_SLOT ?? getNetworkDencunForkSlot(env.NETWORK_NAME),
});

statsSyncer.run({
  cronPatterns: {
    daily: env.STATS_SYNCER_DAILY_CRON_PATTERN,
    overall: env.STATS_SYNCER_OVERALL_CRON_PATTERN,
  },
});

const app = express();

app.use(cors());
app.use(bodyParser.json({ limit: "2mb" }));
app.use(morgan("tiny"));

app.get("/metrics", metricsHandler);

// Serve Swagger UI with our OpenAPI schema
app.use("/", swaggerUi.serve);
app.get("/", swaggerUi.setup(openApiDocument));

// Handle incoming OpenAPI requests
app.use(
  "/",
  createOpenApiExpressMiddleware({
    router: appRouter,
    createContext: createTRPCContext({
      scope: "rest-api",
    }),
    onError({ error }) {
      Sentry.captureException(error);

      logger.error(error);
    },
  })
);

const server = app.listen(env.BLOBSCAN_API_PORT, () => {
  logger.info(
    `REST API server started on http://0.0.0.0:${env.BLOBSCAN_API_PORT}`
  );
});

async function gracefulShutdown(signal: string) {
  logger.debug(`Received ${signal}. Shutting down...`);

  await apiGracefulShutdown()
    .finally(() => statsSyncer.close())
    .finally(() => {
      server.close(() => {
        logger.debug("REST API server shut down successfully");
      });
    });
}

// Listen for TERM signal .e.g. kill
process.on("SIGTERM", () => void gracefulShutdown("SIGTERM"));

// Listen for INT signal e.g. Ctrl-C
process.on("SIGINT", () => void gracefulShutdown("SIGINT"));
