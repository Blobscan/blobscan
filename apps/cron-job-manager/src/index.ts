import * as path from "node:path";
import process from "node:process";
import Graceful from "@ladjs/graceful";
import Bree from "bree";

import { configureHttp } from "./http/configure-http";
import { jobs } from "./jobs";

const bree = new Bree({
  /**
   * Always set the root option when doing any type of
   * compiling with bree. This just makes it clearer where
   * bree should resolve the jobs folder from. By default it
   * resolves to the jobs folder relative to where the program
   * is executed.
   */
  root: path.join(__dirname, "jobs"),
  /**
   * We only need the default extension to be "ts"
   * when we are running the app with ts-node - otherwise
   * the compiled-to-js code still needs to use JS
   */
  defaultExtension: process.env.TS_NODE ? "ts" : "js",
  jobs,
});

// handle graceful reloads, pm2 support, and events like SIGHUP, SIGINT, etc.
const graceful = new Graceful({ brees: [bree] });
graceful.listen();

void (async () => {
  configureHttp();

  await bree.start();
})();
