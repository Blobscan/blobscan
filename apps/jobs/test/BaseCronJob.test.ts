import { beforeEach, describe, expect, it, vi } from "vitest";

import { testValidError } from "@blobscan/test";

import { BaseCronJob, CronJobError } from "../src/BaseCronJob";
import type { BaseCronJobConfig } from "../src/BaseCronJob";

class BaseCronJobMock extends BaseCronJob {
  constructor({ name, jobFn }: Partial<BaseCronJobConfig> = {}) {
    super({
      name: name ?? "test-updater",
      redisUriOrConnection: "redis://localhost:6379/1",
      cronPattern: "* * * * *",
      jobFn: jobFn ?? (() => Promise.resolve()),
    });
  }

  getWorker() {
    if (!this.worker) throw new Error("Worker not initialized");

    return this.worker;
  }

  getQueue() {
    if (!this.queue) throw new Error("Queue not initialized");

    return this.queue;
  }
}

describe("BaseCronJob", () => {
  let cronJob: BaseCronJobMock;

  beforeEach(() => {
    cronJob = new BaseCronJobMock();

    return async () => {
      await cronJob.close();
    };
  });

  it("should create an updater correctly", async () => {
    const queue = cronJob.getQueue();
    const worker = cronJob.getWorker();
    const isPaused = await queue.isPaused();

    expect(worker.isRunning(), "Expected worker to be running").toBeTruthy();
    expect(isPaused, "Expected queue to be running").toBeFalsy();
  });

  describe("when running an updater", () => {
    it("should set up a repeatable job correctly", async () => {
      const queue = cronJob.getQueue();

      await cronJob.start();

      const jobs = await queue.getRepeatableJobs();

      expect(jobs.length, "Expected one repeatable job").toBe(1);
      expect(jobs[0]?.pattern, "Repetable job cron pattern mismatch").toEqual(
        "* * * * *"
      );
    });

    testValidError(
      "should throw a valid error when failing to run",
      async () => {
        const queue = cronJob.getQueue();

        vi.spyOn(queue, "add").mockRejectedValueOnce(new Error("Queue error"));

        await cronJob.start();
      },
      CronJobError,
      { checkCause: true }
    );
  });

  describe("when closing an updater", () => {
    it("should close correctly", async () => {
      const closingCronJob = new BaseCronJobMock();

      await closingCronJob.start();

      const queue = closingCronJob.getQueue();
      const worker = closingCronJob.getWorker();

      const queueCloseSpy = vi.spyOn(queue, "close").mockResolvedValueOnce();
      const queueRemoveAllListenersSpy = vi
        .spyOn(queue, "removeAllListeners")
        .mockReturnValueOnce(queue);

      const workerCloseSpy = vi.spyOn(worker, "close").mockResolvedValueOnce();
      const workerRemoveAllListenersSpy = vi
        .spyOn(worker, "removeAllListeners")
        .mockReturnValueOnce(worker);

      await closingCronJob.close();

      expect(queueCloseSpy).toHaveBeenCalledOnce();
      expect(workerCloseSpy).toHaveBeenCalledOnce();

      expect(queueRemoveAllListenersSpy).toHaveBeenCalledOnce();
      expect(workerRemoveAllListenersSpy).toHaveBeenCalledOnce();
    });
  });

  testValidError(
    "should throw a valid error when failing to close it",
    async () => {
      const queue = cronJob.getQueue();
      const worker = cronJob.getWorker();

      vi.spyOn(queue, "close").mockRejectedValueOnce(
        new Error("Queue closing error")
      );
      vi.spyOn(worker, "close").mockRejectedValueOnce(
        new Error("Worker closing error")
      );

      await cronJob.close();
    },
    CronJobError,
    {
      checkCause: true,
    }
  );
});
