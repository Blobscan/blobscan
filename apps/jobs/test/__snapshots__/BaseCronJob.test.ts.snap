// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseCronJob > should throw a valid error when failing to close it 1`] = `"Cron job \\"test-updater-cron-job\\" failed: An error ocurred when performing closing operation"`;

exports[`BaseCronJob > should throw a valid error when failing to close it 2`] = `[Error: Queue closing error]`;

exports[`BaseCronJob > when running an updater > should throw a valid error when failing to run 1`] = `"Cron job \\"test-updater-cron-job\\" failed: An error ocurred when starting cron job"`;

exports[`BaseCronJob > when running an updater > should throw a valid error when failing to run 2`] = `[Error: Queue error]`;
